lib/mongo/collection/view/readable.rb

```diff
@@ -133,15 +133,15 @@ def count(opts = {})
           cmd[:limit] = opts[:limit] if opts[:limit]
           cmd[:maxTimeMS] = opts[:max_time_ms] if opts[:max_time_ms]
           cmd[:readConcern] = collection.read_concern if collection.read_concern
-          preference = ServerSelector.get(opts[:read] || read)
+          read_pref = opts[:read] || read_preference
           read_with_retry do
-            server = preference.select_server(cluster, false)
+            server = ServerSelector.get(read_pref || server_selector).select_server(cluster, false)
```

Comment:
Good idea. The only code we want to retry is the server selection, not the instantiation of a server selector. I'll change it, as well as in #distinct.